test/integration/apiserver/apiserver_test.go

```diff
@@ -403,6 +406,361 @@ func TestNameInFieldSelector(t *testing.T) {
+				if item.Annotations["foo"] != "baz" {
+					t.Fatalf("unexpected object: %#v", item)
+				}
+
+				if err := client.Namespace(ns).Delete("test-1", &metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &item.UID}}); err != nil {
+					t.Fatal(err)
+				}
+				if _, err := client.Namespace(ns).Get("test-1", metav1.GetOptions{}); !apierrors.IsNotFound(err) {
+					t.Fatal(err)
+				}
+			},
+		},
+		{
+			name: "watch via metadata client",
+			want: func(t *testing.T) {
+				ns := "default"
+				svc, err := clientset.CoreV1().Services(ns).Create(&v1.Service{ObjectMeta: metav1.ObjectMeta{Name: "test-2", Annotations: map[string]string{"foo": "bar"}}, Spec: v1.ServiceSpec{Ports: []v1.ServicePort{{Port: 1000}}}})
+				if err != nil {
+					t.Fatalf("unable to create service: %v", err)
+				}
+				if _, err := clientset.CoreV1().Services(ns).Patch("test-2", types.MergePatchType, []byte(`{"metadata":{"annotations":{"test":"1"}}}`)); err != nil {
+					t.Fatalf("unable to patch cr: %v", err)
+				}
+
+				cfg := metadata.ConfigFor(&restclient.Config{Host: s.URL})
+				wrapper := &callWrapper{}
+				cfg.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+					wrapper.nested = rt
+					return wrapper
+				})
+
+				client := metadata.NewConfigOrDie(cfg).Resource(v1.SchemeGroupVersion.WithResource("services"))
+				w, err := client.Namespace(ns).Watch(metav1.ListOptions{ResourceVersion: svc.ResourceVersion, Watch: true})
+				if err != nil {
+					t.Fatal(err)
+				}
+				defer w.Stop()
+				var r watch.Event
+				select {
+				case evt, ok := <-w.ResultChan():
+					if !ok {
+						t.Fatal("watch closed")
+					}
+					r = evt
+				case <-time.After(5 * time.Second):
+					t.Fatal("no watch event in 5 seconds, bug")
+				}
+				if r.Type != watch.Modified {
+					t.Fatalf("unexpected watch: %#v", r)
+				}
+				item, ok := r.Object.(*metav1.PartialObjectMetadata)
+				if !ok {
+					t.Fatalf("unexpected object: %T", item)
+				}
+				if item.ResourceVersion == "" {
+					t.Fatalf("unexpected object: %#v", item)
+				}
+				if item.Name != "test-2" {
+					t.Fatalf("unexpected object: %#v", item)
+				}
+				if item.UID != svc.UID || item.Annotations["test"] != "1" {
+					t.Fatalf("unexpected object: %#v", item)
+				}
+
+				if wrapper.resp == nil || wrapper.resp.Header.Get("Content-Type") != "application/vnd.kubernetes.protobuf;stream=watch" {
+					t.Fatalf("unexpected response: %#v", wrapper.resp)
+				}
+			},
+		},
+
+		{
+			name: "watch via metadata client on a CRD",
+			want: func(t *testing.T) {
+				ns := "default"
+
+				cr, err := crclient.Create(&unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"apiVersion": "cr.bar.com/v1",
+						"kind":       "Foo",
+						"spec":       map[string]interface{}{"field": 1},
+						"metadata": map[string]interface{}{
+							"name": "test-2",
+							"annotations": map[string]interface{}{
+								"foo": "bar",
+							},
+						},
+					},
+				}, metav1.CreateOptions{})
```

Comment:
Here too I think it would be best if the evaluation half of the test was shared code :/